src/system/SystemPoolStatistics.h

```diff
@@ -0,0 +1,77 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPConfig.h>
+
+#include <algorithm>
+#include <atomic>
+
+#include <system/SystemStats.h>
+
+namespace chip {
+namespace System {
+
+#if CHIP_SYSTEM_CONFIG_PROVIDE_STATISTICS
+
+class ObjectPoolStatistics
+{
+public:
+    void ResetStatistics() { mHighWatermark.store(mUsage.load()); }
+
+    void GetStatistics(Stats::count_t & aNumInUse, Stats::count_t & aHighWatermark)
+    {
+        aNumInUse = static_cast<Stats::count_t>(std::min<unsigned int>(Stats::CHIP_SYS_STATS_COUNT_MAX, mUsage.load()));
+        aHighWatermark =
+            static_cast<Stats::count_t>(std::min<unsigned int>(Stats::CHIP_SYS_STATS_COUNT_MAX, mHighWatermark.load()));
+    }
+
+protected:
+    void IncreaseUsage()
+    {
+        unsigned int usage = ++mUsage;
+        unsigned int prev  = mHighWatermark;
+        while (prev < usage && !mHighWatermark.compare_exchange_weak(prev, usage))
```

Comment:
@kghost , 

I'm having trouble following how this works.  Can you help me understand?

Reply:
When multiple threads are calling `IncreaseUsage()`, one of the thread will grab the highest value of  `usage`, This while loop guarantees that the highest value is set to `mHighWatermark`

https://stackoverflow.com/questions/16190078/how-to-atomically-update-a-maximum-value